### PR TITLE
Cart

### PR DESCRIPTION
**Pull request by alon6699**
Create full working cart and live updates using socket io.
You can check out the client amaRomachClient even though his code isn't 100% clear.

### CODE REVIEW

File: src/routes/product.middleware.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import { Product } from "../models/product.model";
+import { webSocket } from "..";
+import { calculateProductLimit } from "../socket/socket";
+
+const updateProductLimit = (product: Product): Product | Product[] =>
+    product.limit ?
+        ({ ...product.toObject(), limit: product.limit - calculateProductLimit(product.id) }) :
+        product;
+
+export const changeProductClientLimitMiddleware = async (ctx: Context, next: Next): Promise<void> => {
```
Comment (tomporat247) [inline src/routes/product.middleware.ts:13]: rename according to updateProductLimit function

File: src/routes/product.routes.middleware.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import { createProduct, deleteProduct, findProduct, getProducts, updateProduct } from "../controllers/product.controller";
+import { validateProductMiddleware, validatePartialProductMiddleware } from "../middleware/validation/product.validation.middleware";
+import { validateDatabaseConnectionMiddleware } from "../middleware/database/database.middleware";
+import { changeProductClientLimitMiddleware, productChangesMiddleware } from "./product.middleware";
+
+export const productRoutes: Router = new Router({
+    prefix: '/products'
+});
+
+productRoutes.use(validateDatabaseConnectionMiddleware());
```
Comment (tomporat247) [inline src/routes/product.routes.middleware.ts:12]: do not call the middleware fucntion (do it like all others)

File: src/routes/product.routes.middleware.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+productRoutes.get("/", getProducts, changeProductClientLimitMiddleware);
+productRoutes.get("/:id", findProduct, changeProductClientLimitMiddleware);
+productRoutes.post("/", validateProductMiddleware, createProduct, productChangesMiddleware(false));
```
Comment (tomporat247) [inline src/routes/product.routes.middleware.ts:16]: call productChangesMiddleware only in delete

File: src/socket/socket.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
+import { Socket } from "socket.io";
```
Comment (tomporat247) [inline src/socket/socket.ts:1]: rename to websocket, create an http folder as well

File: src/socket/socket.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+import { webSocket } from "..";
+import { Product } from "../models/product.model";
+import { findProductQuery, checkoutQuery } from "../database/product.queries";
+import { logger } from "../logger/logger";
+
+const carts: Record<string, Record<string, number>> = {};
```
Comment (tomporat247) [inline src/socket/socket.ts:8]: manage cart in a different file

File: src/socket/socket.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+const updateProductCartAmount = (socketId: string, product: Product, amount: number) => {
+    amount === 0 ? delete carts[socketId][product._id] : carts[socketId][product._id] = amount;
+    if (product.limit !== undefined) {
+        const limit: number = product.limit - calculateProductLimit(product._id);
+        webSocket.emit('productChanges', { ...product.toObject(), limit });
+    }
+}
+
+export const manageProductInCart = (socket: Socket) => {
```
Comment (tomporat247) [inline src/socket/socket.ts:38]: add return type

File: src/socket/socket.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        .map((id) => manageProductInCart(socket)({ id, amount: 0 })))
+
+export const removeSocket = (socket: Socket) => async () => {
+    logger.info(`A user disconnected ${socket.id}`);
+    await clearCart(socket);
+    delete carts[socket.id];
+}
+
+export const calculateProductLimit = (id: string): number =>
+    Object.values(carts).reduce((acc, productInCart) =>
```
Comment (tomporat247) [inline src/socket/socket.ts:27]: add types
rename productInCart to cart

File: src/socket/socket.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+            }
+        } else {
+            logger.error(`Received negative amount ${amount} to set product ${id} cart amount`);
+        }
+    }
+}
+
+export const checkout = (socket: Socket) => () =>
+    checkoutQuery(carts[socket.id])
+        .then(() => socket.emit('checkout', { error: null }))
```
Comment (tomporat247) [inline src/socket/socket.ts:55]: return {purchased: boolean, error?: string | something else}

File: src/database/product.queries.ts [hunk 9]
```diff
@@ -1,17 +1,50 @@
...
+
+export const checkoutQuery = async (cart: Record<string, number>) => {
+    const session = await startSession();
+    try {
+        session.startTransaction();
+        await Promise.all(Object.keys(cart).map(async (id: string) => {
+            const product: Product = await findProductQuery(id, null, { session: session });
+            if (product.limit) {
+                product.limit -= cart[id];
+                const validation = product.validateSync();
```
[the remark this answers is not in the document]
Comment (alon6699) [reply src/database/product.queries.ts:29]: save() makes validation .. so I used is instead

File: src/routes/product.routes.middleware.ts (hunk 3 above)
Comment (alon6699) [reply src/routes/product.routes.middleware.ts:16]: noo :)

File: src/socket/socket.ts (hunk 6 above)
Comment (alon6699) [reply src/socket/socket.ts:38]: shloud I? its just coping the line below like:
const manageProductInCart = (socket: Socket): ({ id, amount }: { id: string, amount: number }) => void => {
    return async ({ id, amount }: { id: string, amount: number }) => { } }

File: config/env/config.json [hunk 10]
```diff
@@ -1,18 +1,19 @@
 {
-    "port": 3000,
```
Comment (tomporat247) [inline config/env/config.json:2]: folder structure:
divide middlewares to http and websocket

**Pull request merged**